beacon-chain/powchain/block_reader.go

```diff
@@ -84,41 +94,115 @@ func (s *Service) BlockTimeByHeight(ctx context.Context, height *big.Int) (uint6
 }
 
 // BlockNumberByTimestamp returns the most recent block number up to a given timestamp.
-// This is a naive implementation that will use O(ETH1_FOLLOW_DISTANCE) calls to cache
-// or ETH1. This is called for multiple times but only changes every
-// SlotsPerEth1VotingPeriod (1024 slots) so the whole method should be cached.
+// This is an optimized version with the worst case being O(2*repeatedSearches) number of calls
+// while in best case search for the block is performed in O(1).
 func (s *Service) BlockNumberByTimestamp(ctx context.Context, time uint64) (*big.Int, error) {
 	ctx, span := trace.StartSpan(ctx, "beacon-chain.web3service.BlockByTimestamp")
 	defer span.End()
 
-	head, err := s.eth1DataFetcher.HeaderByNumber(ctx, nil)
-	if err != nil {
-		return nil, err
+	if time > s.latestEth1Data.BlockTime {
+		return nil, errors.New("provided time is later than the current eth1 head")
+	}
+	headNumber := big.NewInt(int64(s.latestEth1Data.BlockHeight))
+	headTime := s.latestEth1Data.BlockTime
+	numOfBlocks := uint64(0)
+	estimatedBlk := headNumber.Uint64()
+	maxTimeBuffer := searchThreshold * params.BeaconConfig().SecondsPerETH1Block
+	// Terminate if we cant find an acceptible block after
+	// maxBuffer searches.
+	for i := 0; i < repeatedSearches; i++ {
+		if time > headTime+maxTimeBuffer {
```

Comment:
time> headtime+x ? 
i think it shouldnt be possible with  `if time > s.latestEth1Data.BlockTime {`

Reply:
if you look in the loop we reset the head time each time, so this is to handle those cases.

Reply:
i see `headTime = hinfo.Time`

Reply:
nm